File: src/main/kotlin/specscript/commands/userinteraction/Confirm.kt
package specscript.commands.userinteraction

import specscript.commands.testing.getAnswers
import specscript.language.CommandHandler
import specscript.language.ScriptContext
import specscript.language.SpecScriptCommandError
import specscript.language.ValueHandler
import specscript.language.types.ParameterData
import specscript.util.toDisplayYaml
import tools.jackson.databind.JsonNode
import tools.jackson.databind.node.StringNode
import tools.jackson.databind.node.ValueNode

/**
 * Asks user to confirm a question
 */
object Confirm : CommandHandler("Confirm", "core/user-interaction"), ValueHandler {

    val yes = StringNode("Yes")
    val no = StringNode("No")

    override fun execute(data: ValueNode, context: ScriptContext): JsonNode? {

        val question = data.toDisplayYaml()

        // Skip confirmation if running in non-interactive mode
        if (!context.interactive) {
            return null
        }

        val confirmationDialog = ParameterData(
            description = question,
            enum = listOf(yes, no)
        )

        val answer = confirmationDialog.ask(confirmationDialog.question(), context.getAnswers(), context.interactive)

        if (answer == no) {
            throw SpecScriptCommandError("No confirmation -- action canceled.")
        }

        return null
    }
}

File: typescript/src/commands/confirm.ts
/**
 * Confirm command: asks the user for yes/no confirmation.
 *
 * Presents a Yes/No choice. Returns "Yes" on confirmation.
 * Throws SpecScriptCommandError on rejection ("No").
 */

import type { CommandHandler } from '../language/command-handler.js'
import type { ScriptContext } from '../language/context.js'
import type { JsonValue } from '../language/types.js'
import { isString, SpecScriptCommandError } from '../language/types.js'
import { promptSelect, getAnswers } from '../language/user-prompt.js'
import type { Choice } from '../language/user-prompt.js'

export const ConfirmCommand: CommandHandler = {
  name: 'Confirm',

  async execute(data: JsonValue, context: ScriptContext): Promise<JsonValue | undefined> {
    if (!isString(data)) return undefined

    const message = data
    const choices: Choice[] = [
      { displayName: 'Yes', value: 'Yes' },
      { displayName: 'No', value: 'No' },
    ]

    const answers = getAnswers(context.session)
    const stdout = context.session.get('stdout') as ((text: string) => void) | undefined

    const answer = await promptSelect(answers, message, choices, false, stdout, context.interactive)

    if (answer === 'No') {
      throw new SpecScriptCommandError(
        'No confirmation -- action canceled.',
        'Confirmation',
      )
    }

    return answer
  },
}
